src/components/TodoList.tsx

```diff
@@ -0,0 +1,214 @@
+import classNames from 'classnames';
+import { useEffect, useMemo, useState } from 'react';
+import { useTodoContext } from '../context';
+import { Todo } from '../types/Todo';
+import { deleteTodo } from '../api/todos';
+import { Errors } from '../types/Errors';
+
+type Props = {
+  tempTodo: Todo | null
+};
+
+export const TodoList = ({ tempTodo }: Props) => {
+  const {
+    activeFilter,
+    handleTodosFilter,
+    visibleTodos,
+    setVisibleTodos,
+    allTodos,
+    errorHandler,
+    setAllTodos,
+    inputRef,
+  } = useTodoContext();
+
+  const [isLoading, setIsLoading] = useState<number | boolean>(false);
+
+  const handleTodoDelete = (id: number) => {
+    setIsLoading(id);
+
+    const deletingTodo = async () => {
+      try {
+        if (allTodos) {
+          await deleteTodo(id);
+          const updatedTodo = allTodos.filter(todo => todo.id !== id);
+
+          setAllTodos(updatedTodo);
+        }
+      } catch (error) {
+        errorHandler(Errors.deleteError);
+      } finally {
+        setIsLoading(false);
+      }
+    };
+
+    deletingTodo();
+  };
+
+  const handleClearCompleted = async () => {
+    if (allTodos) {
+      const completedTodos = allTodos.filter(todo => todo.completed);
+
+      try {
+        await Promise.all(completedTodos.map(todo => deleteTodo(todo.id)));
+        setAllTodos(allTodos.filter(todo => !todo.completed));
+      } catch (error) {
+        errorHandler(Errors.deleteError);
+      }
+    }
+  };
+
+  const activeTodos = useMemo(() => {
+    let counter = 0;
+
+    allTodos?.forEach(todo => {
+      if (!todo.completed) {
+        counter += 1;
+      }
+    });
+
+    return counter;
+  }, [allTodos]);
+
+  useEffect(() => {
+    setVisibleTodos(allTodos);
+  }, [allTodos, setVisibleTodos]);
+
+  useEffect(() => {
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [allTodos, inputRef]);
+
+  return (
+    <>
+      {visibleTodos?.map(todo => (
```

Comment:
I think you should create a component for a todo, maybe you could use it for tempTodo either. This would make `TodoList` way cleaner